conda/core/index.py

```diff
@@ -35,17 +35,30 @@ def check_allowlist(channel_urls: list[str]) -> None:
 
     :param channel_urls: A list of channel URLs to check against the allowlist.
     :raises ChannelNotAllowed: If any URL is not in the allowlist.
+    :raises ChannelDenied: If any URL is in the denylist.
     """
-    if context.allowlist_channels:
-        allowlist_channel_urls = tuple(
-            chain.from_iterable(
-                Channel(c).base_urls for c in context.allowlist_channels
-            )
+    allowlist_channel_urls = tuple(
+        chain.from_iterable(
+            Channel(allowlist_channel).base_urls
+            for allowlist_channel in context.allowlist_channels
+        )
+    )
+    denylist_channel_urls = tuple(
+        chain.from_iterable(
+            Channel(denylist_channel).base_urls
+            for denylist_channel in context.denylist_channels
         )
-        for url in channel_urls:
-            these_urls = Channel(url).base_urls
-            if not all(this_url in allowlist_channel_urls for this_url in these_urls):
-                raise ChannelNotAllowed(Channel(url))
+    )
+    for channel_url in channel_urls:
+        channel = Channel(channel_url)
+        for channel_base_url in channel.base_urls:
+            if channel_base_url in denylist_channel_urls:
+                raise ChannelDenied(channel)
+            if (
+                context.allowlist_channels
+                and channel_base_url not in allowlist_channel_urls
+            ):
+                raise ChannelNotAllowed(channel)
```

Comment:
two edits:
 1. only iterate over channel_urls if either allowlist or denylist are defined
 2. use `allowlist_channel_urls` instead of `context.allowlist_channels`

```suggestion
    if allowlist_channel_urls or denylist_channel_urls:
        for channel_url in channel_urls:
            channel = Channel(channel_url)
            for channel_base_url in channel.base_urls:
                if channel_base_url in denylist_channel_urls:
                    raise ChannelDenied(channel)
                if (
                    allowlist_channel_urls
                    and channel_base_url not in allowlist_channel_urls
                ):
                    raise ChannelNotAllowed(channel)
```